bpo-32388: Remove cross-version binary compatibility requirement in tp_flags

pitrou opened the pull request:
Starting with 3.8, it is now allowed to add new fields at the end of the `PyTypeObject` struct in feature releases without having to allocate a dedicated compatibility flag in `tp_flags`. This will reduce the risk of running out of bits in the 32-bit `tp_flags` value.

<!-- issue-number: [bpo-32388](https://bugs.python.org/issue32388) -->
https://bugs.python.org/issue32388
<!-- /issue-number -->

scoder commented on Include/object.h:
```diff
@@ -675,8 +672,14 @@ given type object has a specified feature.
  * Python 3.0 transition). */
 
+/* The following flags are kept for compatibility, but all types are
+ * supposed to support the given functionality.  Starting with 3.7,
```
This explicitly mentions the CPython release of the change and needs adaptation when this gets merged.

JimJJewett replied:
I think it should also be mentioned that Py_TPFLAGS_HAVE_FINALIZE is being rolled into the always-supported default, if that effects how to write objects.  If it does not affect anything extension writers do, then that (and the reason it doesn't) should be mentioned.

JimJJewett replied:
Conversely, if it does affect anything for extension writers -- even just simplifying things -- that should be added to the porting section of what's new.

JimJJewett commented on Objects/object.c:
```diff
@@ -276,8 +276,7 @@ PyObject_CallFinalizer(PyObject *self)
 
     /* The former could happen on heaptypes created from the C API, e.g.
        PyType_FromSpec(). */
```
I think this comment (including the existence of a "former" depends on removed code, and should therefore itself be removed.

pitrou committed:
bpo-32388: Remove cross-version binary compatibility requirement in tp_flags

It is now allowed to add new fields at the end of the PyTypeObject struct
(and even in the middle, if we are so inclined) in feature releases without
having to allocate a dedicated compatibility flag in tp_flags.

This will reduce the risk of running out of bits in the 32-bit tp_flags value.

tirkarthi commented on Doc/whatsnew/3.8.rst:
```diff
@@ -1019,6 +1019,13 @@ Changes in the C API
+  exported by a third-party extension module is supposed to have all the
+  slots expected in the current Python version, including
+  :c:member:`~PyTypeObject.tp_finalize` (:const:`Py_TPFLAGS_HAVE_FINALIZE`
+  is not checked anymore before reading :c:member:`~PyTypeObject.tp_finalize`).
```
Could probably link to bpo with :issue:`32388` since it's in What's New page.

jdemeyer commented on Include/object.h:
```diff
@@ -332,8 +329,14 @@ given type object has a specified feature.
 /* NOTE: The following flags reuse lower bits (removed as part of the
  * Python 3.0 transition). */
 
+/* The following flags are kept for compatibility, but all types are
```
`flags are` -> `flag is`

pitrou committed:
Address review comments

pitrou committed:
Merge branch 'master' into no_tpflags_compatibility

pitrou committed:
Restore previous tp_flag value
